rfc-149-switch-to-per-page-asset-loading.md

```diff
@@ -0,0 +1,208 @@
+
+![](rfc-149/css-size-for-top-100-most-visited-pages-on-govuk.svg)
+
+
+This means that all visitors will see a smaller page size, regardless of whether they visit one page or multiple pages. Six out of ten tracked visitors to GOVUK only view a single page.
+
+Concatenating the assets can provide a benefit for those that visit multiple pages - but this doesn't work on GOVUK due to the number of rendering apps used and the current lack of a shared space to put assets.
+
+### HTTP requests
+
+The number of requests per page would go up - this shouldn't be a problem as [GOV.UK has HTTP/2 enabled][http2], and is due to have [HTTP/3][http3] enabled in the near future. During the past month 99.5% of tracked visitors used browsers that support HTTP/2, and 66.2% of visitors used browsers that supported HTTP/3.
+
+Of the top 100 most visited pages on GOV.UK, the average number of components on a page is 19. The smallest number of components on a page is 9, and the largest number of components is 25. It's worth noting that most - but not all - components need a stylesheet (72 out of 78 components have a stylesheet) and not every component needs JavaScript (24 out of 78 components have a JavaScript file) - so adding a component to a page doesn't always mean two more requests are made.
+
+It's worth noting that this change will lead to slower load times for browsers that only support HTTP/1.1 - even when considering the reduction in page size. ([More information on HTTP/1.1's limitations][http_limitations]). This will impact Internet Explorer 11 on Windows 8 and below; the first versions of Chrome, Firefox, and Opera that supported HTTP/2 were released in early 2015.
+
+This balance between making the site load faster for a majority of users is not just a numbers game - the users who are most likely to be using older browsers are those that are likely to least afford the upgrade cost for a newer device, and they're likely to be the users who most need the information on GOV.UK.
+
+Whilst this proposed change will make things slower for users whose browser doesn't support HTTP/2 or HTTP/3, it won't prevent pages from loading. This proposed change should be accepted because it will dramatically improve page performance for almost all visitors without blocking access for a small number of visitors.
+
+### Non-rendering applications that use the components gem
+
+There are applications that use the components gem, but don't render public pages on GOV.UK - for example, publishing applications and the Govspeak gem. These may have a different set up with different ways of rendering pages.
+
+These changes should be optional - even if this is the default behaviour, it should be able to be turned off at a component by component level and at an application level. This could be done with a helper method that checks both if the component has already been used, and if adding the component assets are allowed:
```

Comment:
This adds a fair bit of complexity. Is there an advantage to making things optional?
Do any applications use govuk_publishing_components without Static / Slimmer?

Reply:
The publishing applications and the govspeak gem use the components gem without Static - I'll made it clearer in the previous paragraph.

Reply:
Added in https://github.com/alphagov/govuk-rfcs/pull/152/commits/64480aac48b0feff8489c7036525c546a4b079ef